Source/CBLJSFunction.m

```diff
@@ -48,121 +43,82 @@ - (instancetype) init {
 }
 
 
-- (void)dealloc {
-    if (_context)
-        JSGlobalContextRelease(_context);
-}
-
-
 @end
 
 
 
+#define kCBLJSFunctionName @"cblJSFunction"
 
-@implementation CBLJSFunction
-{
+@implementation CBLJSFunction {
     CBLJSCompiler* _compiler;
-    unsigned _nParams;
-    JSObjectRef _fn;
+    NSString *_fnName;
 }
 
-- (instancetype) initWithCompiler: (CBLJSCompiler*)compiler
-                       sourceCode: (NSString*)source
-                       paramNames: (NSArray*)paramNames
-{
+
+- (instancetype) initWithCompiler: (CBLJSCompiler*)compiler sourceCode: (NSString*)source {
     self = [super init];
     if (self) {
         _compiler = compiler;
-        _nParams = (unsigned)paramNames.count;
-
-        // The source code given is a complete function, like "function(doc){....}".
-        // But JSObjectMakeFunction wants the source code of the _body_ of a function.
-        // Therefore we wrap the given source in an expression that will call it:
-        NSString* body = [NSString stringWithFormat: @"return (%@)(%@);",
-                                               source, [paramNames componentsJoinedByString: @","]];
-
-        // Compile the function:
-        JSStringRef jsParamNames[_nParams];
-        for (NSUInteger i = 0; i < _nParams; ++i)
-            jsParamNames[i] = JSStringCreateWithCFString((__bridge CFStringRef)paramNames[i]);
-        JSStringRef jsBody = JSStringCreateWithCFString((__bridge CFStringRef)body);
-        JSValueRef exception;
-        _fn = JSObjectMakeFunction(_compiler.context, NULL, _nParams, jsParamNames, jsBody,
-                                   NULL, 1, &exception);
-        JSStringRelease(jsBody);
-        for (NSUInteger i = 0; i < _nParams; ++i)
-            JSStringRelease(jsParamNames[i]);
-        
-        if (!_fn) {
-            WarnJSException(_compiler.context, @"JS function compile failed", exception);
+
+        // Evaluate source code:
+        _fnName = [self functionNameFromSourceCode: source];
+        NSString* script = [NSString stringWithFormat: @"var %@ = %@", _fnName, source];
```

Comment:
I don't think you need to assign the function to a named variable in JS — if you evaluate the source, the result will be the JS function object. Then you can call that directly in the -call method instead of looking up a variable from the context.